src/gt4py/next/backend.py

```diff
@@ -26,143 +27,138 @@
-def foast_to_foast_closure(
-    inp: workflow.InputWithArgs[ffront_stages.FoastOperatorDefinition],
-) -> ffront_stages.FoastClosure:
-    from_fieldop = inp.kwargs.pop("from_fieldop")
-    return ffront_stages.FoastClosure(
-        foast_op_def=inp.data,
-        args=inp.args,
-        kwargs=inp.kwargs,
-        closure_vars={inp.data.foast_node.id: from_fieldop},
-    )
+from gt4py.next.otf import arguments, stages, workflow
+from gt4py.next.program_processors import modular_executor
 
 
-@dataclasses.dataclass(frozen=True)
-class FieldopTransformWorkflow(workflow.NamedStepSequenceWithArgs):
-    """Modular workflow for transformations with access to intermediates."""
-
-    func_to_foast: workflow.SkippableStep[
-        ffront_stages.FieldOperatorDefinition | ffront_stages.FoastOperatorDefinition,
-        ffront_stages.FoastOperatorDefinition,
-    ] = dataclasses.field(
-        default_factory=lambda: func_to_foast.OptionalFuncToFoastFactory(cached=True)
-    )
-    foast_to_foast_closure: workflow.Workflow[
-        workflow.InputWithArgs[ffront_stages.FoastOperatorDefinition], ffront_stages.FoastClosure
-    ] = dataclasses.field(default=foast_to_foast_closure, metadata={"takes_args": True})
-    foast_to_past_closure: workflow.Workflow[
-        ffront_stages.FoastClosure, ffront_stages.PastClosure
-    ] = dataclasses.field(
-        default_factory=lambda: foast_to_past.FoastToPastClosure(
-            foast_to_past=workflow.CachedStep(
-                foast_to_past.foast_to_past, hash_function=ffront_stages.fingerprint_stage
-            )
-        )
-    )
-    past_transform_args: workflow.Workflow[ffront_stages.PastClosure, ffront_stages.PastClosure] = (
-        dataclasses.field(default=past_process_args.past_process_args)
-    )
-    past_to_itir: workflow.Workflow[ffront_stages.PastClosure, stages.ProgramCall] = (
-        dataclasses.field(default_factory=past_to_itir.PastToItirFactory)
-    )
+ARGS: typing.TypeAlias = arguments.JITArgs
+CARG: typing.TypeAlias = arguments.CompileTimeArgs
+IT_PRG: typing.TypeAlias = itir.FencilDefinition
 
-    foast_to_itir: workflow.Workflow[ffront_stages.FoastOperatorDefinition, itir.Expr] = (
-        dataclasses.field(
-            default_factory=lambda: workflow.CachedStep(
-                step=foast_to_itir.foast_to_itir, hash_function=ffront_stages.fingerprint_stage
-            )
-        )
-    )
 
-    @property
-    def step_order(self) -> list[str]:
-        return [
-            "func_to_foast",
-            "foast_to_foast_closure",
-            "foast_to_past_closure",
-            "past_transform_args",
-            "past_to_itir",
-        ]
+INPUT_DATA: typing.TypeAlias = DSL_FOP | FOP | DSL_PRG | PRG | IT_PRG
+INPUT_PAIR: typing.TypeAlias = workflow.DataArgsPair[INPUT_DATA, ARGS | CARG]
 
 
-DEFAULT_FIELDOP_TRANSFORMS = FieldopTransformWorkflow()
+@dataclasses.dataclass(frozen=True)
+class Transforms(workflow.MultiWorkflow[INPUT_PAIR, stages.AOTProgram]):
```

Comment:
@tehrengruber Could you please review the new "Transforms" toolchain with a view to the tradeoff between "pluggability" of steps (the ability to combine them without copy-pasting glue code) and what's necessary to achieve it? I believe that in this version the tradeoff is much more favorable than previously.